docs/proposals/nodesets.md

```diff
@@ -0,0 +1,157 @@
+(i.e. it can take account of pods _not_ scheduled).
+
+### Support multiple NodeSets of different instance types
+
+Users may want to run instances of different types in the same cluster, permanently or transiently.
+Permanently mixing instance types may require more scheduler support, but for transient usage (changing
+instance sizes) this is helpful for changing a cluster.
+
+Users may also want to mix instance types in a way that is not visible to the cluster, for example mixing
+on-demand and spot instances on AWS.
+
+### Support upgrading by specifying software versions
+
+By specifying that a new NodeSet should be launched with a different image or kubernetes version,
+users can easily perform an online upgrade of their nodes (within compatible versions).
+
+## Functionality
+
+The NodeSet supports standard API features:
+- create
+  - The spec for NodeSet includes a count of nodes
+  - The spec includes a template, which specifies the type of nodes that will be launched as labels.
+    We imagine that a cloudprovider actually has a pool of machines of various instance types, images,
+    kubernetes versions etc.  We will label each node with the labels describing its instance type, running image,
+    kubernetes version etc on creation.  Then the selector acts to specify the information that nodes must have.
+    Thus, specifying `kubernetes.io/instanceType=n1-highmem-16` will ensure that nodes are launched with
+    `n1-highmem-16` instance type.  Selectors with the `kubernetes.io/` prefix must be known, or it is a
+    validation error; selectors without the prefix are treated as normal labels.
+    (TODO: Is this too 'cute'?)
+  - YAML example:
+
+```YAML
+  apiVersion: v1
+  kind: NodeSet
+  metadata:
+    labels:
+      bornondate: 2015-10-23
+    name: bigmemory-node-pool1
+  spec:
+    count: 4
+    template:
+      metadata:
+        labels:
+          kubernetes.io/instanceType: n1-highmem-16
+```
+
+  - commands that get info
+    - get (e.g. kubectl get nodesets)
+    - describe
+  - Modifiers
+    - delete
+    - label
+	- annotate
+    - update operations like patch and replace (only allowed to NodeSet metadata.labels and spec.count)
+  - In general, for all the supported features like get, describe, update, etc, the NodeSet works in a similar way to the Replication Controller.
+
+
+## Design
+
+#### Client
+
+- Add support for NodeSet commands to kubectl and the client. Client code will be added to the experimental area of client/unversioned.
+
+#### Apiserver
+
+- Accept, parse, validate client commands
+- REST API calls are handled in registry/daemon
+  - In particular, the api server will add the object to etcd
+  - NodeSetManager listens for updates to etcd (using Framework.informer)
+- API objects & validation for NodeSet will be created alongside other experimental types.
+
+#### NodeSetManager
+
+- Creates new nodes and deletes nodes in response to changes as observed.
```

Comment:
I think cluster-autoscaling #15304 is actually very complementary to this proposal, and I imagine combining them.  I think cluster-autoscaling addresses the health-checking / scaling omissions here.  And I think this proposal makes cluster-autoscaling more useful, in that the NodeSet is a place to configure the health-checking/scaling policies in the API.  (Thanks @bgrant0607 for pointing me to #15304!)